lib/dispatcher/env-http-proxy-agent.js

```diff
@@ -0,0 +1,108 @@
+'use strict'
+
+const DispatcherBase = require('./dispatcher-base')
+const { kClose, kDestroy, kClosed, kDestroyed, kDispatch, kNoProxyAgent, kHttpProxyAgent, kHttpsProxyAgent } = require('../core/symbols')
+const ProxyAgent = require('./proxy-agent')
+const Agent = require('./agent')
+
+const DEFAULT_PORTS = {
+  'http:': 80,
+  'https:': 443
+}
+
+class EnvHttpProxyAgent extends DispatcherBase {
+  constructor (opts) {
+    super()
+
+    this[kNoProxyAgent] = new Agent(opts)
+
+    const HTTP_PROXY = process.env.HTTP_PROXY || process.env.http_proxy
+    if (HTTP_PROXY) {
+      this[kHttpProxyAgent] = new ProxyAgent({ ...opts, uri: HTTP_PROXY })
+    } else {
+      this[kHttpProxyAgent] = this[kNoProxyAgent]
+    }
+
+    const HTTPS_PROXY = process.env.HTTPS_PROXY || process.env.https_proxy
+    if (HTTPS_PROXY) {
+      this[kHttpsProxyAgent] = new ProxyAgent({ ...opts, uri: HTTPS_PROXY })
+    } else {
+      this[kHttpsProxyAgent] = this[kHttpProxyAgent]
+    }
+  }
+
+  [kDispatch] (opts, handler) {
+    const url = new URL(opts.origin)
+    const agent = this.#getProxyAgentForUrl(url)
+    return agent.dispatch(opts, handler)
+  }
+
+  async [kClose] () {
+    await this[kNoProxyAgent].close()
+    if (!this[kHttpProxyAgent][kClosed]) {
+      await this[kHttpProxyAgent].close()
+    }
+    if (!this[kHttpsProxyAgent][kClosed]) {
+      await this[kHttpsProxyAgent].close()
+    }
+  }
+
+  async [kDestroy] (err) {
+    await this[kNoProxyAgent].destroy(err)
+    if (!this[kHttpProxyAgent][kDestroyed]) {
+      await this[kHttpProxyAgent].destroy(err)
+    }
+    if (!this[kHttpsProxyAgent][kDestroyed]) {
+      await this[kHttpsProxyAgent].destroy(err)
+    }
+  }
+
+  #getProxyAgentForUrl (url) {
+    let { protocol, host: hostname, port } = url
+
+    // Stripping ports in this way instead of using parsedUrl.hostname to make
+    // sure that the brackets around IPv6 addresses are kept.
+    hostname = hostname.replace(/:\d*$/, '').toLowerCase()
+    port = Number.parseInt(port, 10) || DEFAULT_PORTS[protocol] || 0
+    if (!this.#shouldProxy(hostname, port)) {
+      return this[kNoProxyAgent]
+    }
+    if (protocol === 'https:') {
+      return this[kHttpsProxyAgent]
+    }
+    return this[kHttpProxyAgent]
+  }
+
+  #shouldProxy (hostname, port) {
+    const NO_PROXY = process.env.NO_PROXY || process.env.no_proxy
+    if (!NO_PROXY) {
+      return true // Always proxy if NO_PROXY is not set.
+    }
+    if (NO_PROXY === '*') {
+      return false // Never proxy if wildcard is set.
+    }
+
+    return NO_PROXY.split(/[,\s]/).filter((entry) => !!entry.length).every(function (entry) {
```

Comment:
Why would you always split NO_PROXY every time. You should pre compute it.

Reply:
env vars can be mutated.

```js
await fetch('...')
process.env.NO_PROXY = '*'
await fetch('...')
```

For performance reasons it should be possible to compare old/new values and only parse when a new one is set.

Reply:
this should also probably just be a for loop instead of .filter().every()

Reply:
I don't see a reason that it shouldn't update when changed